wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,1093 @@
+		}
+
+		$ast = $children[0]->get_child_node();
+		switch ( $ast->rule_name ) {
+			case 'selectStatement':
+				$this->query_type = 'SELECT';
+				$query            = $this->translate( $ast->get_child() );
+				$stmt             = $this->execute_sqlite_query( $query );
+				$this->set_results_from_fetched_data(
+					$stmt->fetchAll( $this->pdo_fetch_mode )
+				);
+				break;
+			case 'insertStatement':
+			case 'updateStatement':
+				$this->execute_update_statement( $ast );
+				break;
+			case 'replaceStatement':
+			case 'deleteStatement':
+				if ( 'insertStatement' === $ast->rule_name ) {
+					$this->query_type = 'INSERT';
+				} elseif ( 'replaceStatement' === $ast->rule_name ) {
+					$this->query_type = 'REPLACE';
+				} elseif ( 'deleteStatement' === $ast->rule_name ) {
+					$this->query_type = 'DELETE';
+				}
+				$query = $this->translate( $ast );
+				$this->execute_sqlite_query( $query );
+				$this->set_result_from_affected_rows();
+				break;
+			case 'createStatement':
+				$this->query_type = 'CREATE';
+				$subtree          = $ast->get_child_node();
+				switch ( $subtree->rule_name ) {
+					case 'createTable':
+						$query = $this->translate( $ast );
+						$this->execute_sqlite_query( $query );
+						$this->set_result_from_affected_rows();
+						break;
+					default:
+						throw new Exception(
+							sprintf(
+								'Unsupported statement type: "%s" > "%s"',
+								$ast->rule_name,
+								$subtree->rule_name
+							)
+						);
+				}
+				break;
+			default:
+				throw new Exception( sprintf( 'Unsupported statement type: "%s"', $ast->rule_name ) );
+		}
+	}
+
+	private function execute_update_statement( WP_Parser_Node $node ): void {
+		// @TODO: Add support for UPDATE with multiple tables and JOINs.
+		//        SQLite supports them in the FROM clause.
+
+		$has_order = $node->has_child_node( 'orderClause' );
+		$has_limit = $node->has_child_node( 'simpleLimitClause' );
+
+		/*
+		 * SQLite doesn't support UPDATE with ORDER BY/LIMIT.
+		 * We need to use a subquery to emulate this behavior.
+		 *
+		 * For instance, the following query:
+		 *   UPDATE t SET c = 1 WHERE c = 2 LIMIT 1;
+		 * Will be rewritten to:
+		 *   UPDATE t SET c = 1 WHERE rowid IN ( SELECT rowid FROM t WHERE c = 2 LIMIT 1 );
+		 */
+		if ( $has_order || $has_limit ) {
+			$subquery = 'SELECT rowid FROM ' . $this->translate_sequence(
+				array(
+					$node->get_descendant_node( 'tableReference' ),
+					$node->get_descendant_node( 'whereClause' ),
+					$node->get_descendant_node( 'orderClause' ),
+					$node->get_descendant_node( 'simpleLimitClause' ),
+				)
+			);
+
+			$update_nodes = array();
+			foreach ( $node->get_children() as $child ) {
+				$update_nodes[] = $child;
+				if (
+					$child instanceof WP_Parser_Node
+					&& 'updateList' === $child->rule_name
+				) {
+					// Skip WHERE, ORDER BY, and LIMIT.
+					break;
+				}
+			}
+			$query = $this->translate_sequence( $update_nodes )
```

Comment:
Thought: we could have a single translate() method that would support both strings and arrays

Reply:
@adamziel Looking at this simple use case, I agree that it would make sense, but the whole context is that has an additional parameter (and maybe more, in the future) that doesn't really fit into the `translate()` recursion.

Joining the subtrees by space is the default:
```php
private function translate_sequence( array $nodes, string $separator = ' ' ): string
```

But there are other use cases with a different separator, such as a list of column definitions — `translate_sequence( $nodes, ', ' )`, etc. I think we may also need a `$prefix` and `$suffix` in the future, e.g. to wrap the subtrees in parentheses, etc. It really is just a helper so we don't have to repeat the same loop in multiple places.

Reply:
makes sense, thanks for explaining!